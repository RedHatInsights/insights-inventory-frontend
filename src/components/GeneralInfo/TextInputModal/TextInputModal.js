import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  FormHelperText,
  HelperText,
  HelperTextItem,
  Modal,
  TextInput,
  ValidatedOptions,
} from '@patternfly/react-core';

export default class TextInputModal extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  static getDerivedStateFromProps(props, state) {
    if (!props.isOpen) {
      return { value: undefined };
    }

    if (state.value !== undefined) {
      return null;
    }

    return {
      value: props.value || '',
    };
  }

  render() {
    const {
      title,
      isOpen,
      onCancel,
      onSubmit,
      ariaLabel,
      modalOuiaId,
      cancelOuiaId,
      confirmOuiaId,
      inputOuiaId,
    } = this.props;
    const { value } = this.state;

    return (
      <Modal
        variant="small"
        title={title}
        className="ins-c-inventory__detail--edit"
        aria-label={ariaLabel ? `${ariaLabel} - modal` : 'input modal'}
        ouiaId={modalOuiaId}
        isOpen={isOpen}
        onClose={(event) => onCancel(event)}
        actions={[
          <Button
            key="confirm"
            data-action="confirm"
            variant="primary"
            onClick={() => onSubmit(this.state.value?.trim())}
            ouiaId={confirmOuiaId}
            isDisabled={
              this.props.value === this.state.value ||
              this.state.value?.trim().length === 0
            }
          >
            Save
          </Button>,
          <Button
            key="cancel"
            data-action="cancel"
            variant="link"
            onClick={onCancel}
            ouiaId={cancelOuiaId}
          >
            Cancel
          </Button>,
        ]}
      >
        <TextInput
          value={value}
          type="text"
          ouiaId={inputOuiaId}
          onChange={(_event, value) => this.setState({ value })}
          aria-label={ariaLabel}
          validated={value?.trim().length === 0 && ValidatedOptions.error}
        />
        <FormHelperText>
          <HelperText id="helper-text2" aria-live="polite">
            {value?.trim().length === 0 && (
              <HelperTextItem variant="error">
                Name cannot be blank
              </HelperTextItem>
            )}
          </HelperText>
        </FormHelperText>
      </Modal>
    );
  }
}

TextInputModal.propTypes = {
  title: PropTypes.string,
  isOpen: PropTypes.bool,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
  ariaLabel: PropTypes.string,
  modalOuiaId: PropTypes.string,
  cancelOuiaId: PropTypes.string,
  confirmOuiaId: PropTypes.string,
  inputOuiaId: PropTypes.string,
  value: PropTypes.string,
};

TextInputModal.defaultProps = {
  onCancel: () => undefined,
  onSubmit: () => undefined,
  isOpen: false,
  title: '',
  ariaLabel: 'input text',
};
