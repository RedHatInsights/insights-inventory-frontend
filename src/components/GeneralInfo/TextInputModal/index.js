export { default } from './TextInputModal';
export * from './TextInputModal';
