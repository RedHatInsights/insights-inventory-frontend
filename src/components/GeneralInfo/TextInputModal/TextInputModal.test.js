import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import TextInputModal from './TextInputModal';

describe('TextInputModal', () => {
  describe('getDerivedStateFromProps', () => {
    it('should set state value to undefined', () => {
      expect(
        TextInputModal.getDerivedStateFromProps(
          {
            isOpen: false,
            value: 'some-value',
          },
          { value: 'test' }
        )
      ).toEqual({
        value: undefined,
      });
    });

    it('should keep the value same if isOpen set and state value set', () => {
      expect(
        TextInputModal.getDerivedStateFromProps(
          {
            isOpen: true,
            value: 'some-value',
          },
          { value: 'test' }
        )
      ).toBe(null);
    });

    it('should set the state value', () => {
      expect(
        TextInputModal.getDerivedStateFromProps(
          {
            isOpen: true,
            value: 'some-value',
          },
          { value: undefined }
        )
      ).toEqual({
        value: 'some-value',
      });
    });
  });

  describe('render', () => {
    it('should render without any props', () => {
      const view = render(<TextInputModal />);
      expect(view.baseElement).toMatchSnapshot();
    });

    it('should render open', () => {
      const view = render(<TextInputModal isOpen />);
      expect(view.baseElement).toMatchSnapshot();
    });

    it('should render title', () => {
      const view = render(<TextInputModal isOpen title="Some title" />);
      expect(view.baseElement).toMatchSnapshot();
    });

    it('should render aria label', () => {
      const view = render(
        <TextInputModal isOpen ariaLabel="Some aria label" />
      );
      expect(view.baseElement).toMatchSnapshot();
    });
  });

  describe('API', () => {
    it('getDerivedStateFromProps should be called', async () => {
      const getDerivedStateFromProps = jest.spyOn(
        TextInputModal,
        'getDerivedStateFromProps'
      );
      render(<TextInputModal isOpen />);

      await userEvent.type(
        screen.getByRole('textbox', {
          name: /input text/i,
        }),
        'some'
      );
      expect(getDerivedStateFromProps).toBeCalled();
    });

    it('onCancel should be called', async () => {
      const onCancel = jest.fn();
      render(<TextInputModal isOpen onCancel={onCancel} />);

      await userEvent.click(
        screen.getByRole('button', {
          name: /cancel/i,
        })
      );
      expect(onCancel).toBeCalled();
    });

    it('onSubmit should be called', async () => {
      const onSubmit = jest.fn();
      render(<TextInputModal isOpen onSubmit={onSubmit} />);

      await userEvent.click(
        screen.getByRole('button', {
          name: /save/i,
        })
      );
      expect(onSubmit).toBeCalled();
    });

    it('X button should call onClose', async () => {
      const onCancel = jest.fn();
      render(<TextInputModal isOpen onCancel={onCancel} />);

      await userEvent.click(
        screen.getByRole('button', {
          name: /close/i,
        })
      );
      expect(onCancel).toBeCalled();
    });
  });
});
