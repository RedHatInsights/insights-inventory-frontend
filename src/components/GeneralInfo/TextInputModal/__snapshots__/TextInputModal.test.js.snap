// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputModal render should render aria label 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-5"
          aria-label="Some aria label - modal"
          aria-labelledby="pf-modal-part-3"
          aria-modal="true"
          class="pf-c-modal-box ins-c-inventory__detail--edit pf-m-sm"
          data-ouia-component-id="OUIA-Generated-Modal-small-4"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-3"
          role="dialog"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Modal-small-4-ModalBoxCloseButton"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-5"
          >
            <input
              aria-invalid="false"
              aria-label="Some aria label"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-3"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              type="text"
              value=""
            />
          </div>
          <footer
            class="pf-c-modal-box__footer"
          >
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-primary"
              data-action="confirm"
              data-ouia-component-id="OUIA-Generated-Button-primary-3"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Save
            </button>
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link"
              data-action="cancel"
              data-ouia-component-id="OUIA-Generated-Button-link-3"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Cancel
            </button>
          </footer>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TextInputModal render should render open 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-3"
          aria-label="input text - modal"
          aria-labelledby="pf-modal-part-1"
          aria-modal="true"
          class="pf-c-modal-box ins-c-inventory__detail--edit pf-m-sm"
          data-ouia-component-id="OUIA-Generated-Modal-small-2"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-1"
          role="dialog"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Modal-small-2-ModalBoxCloseButton"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-3"
          >
            <input
              aria-invalid="false"
              aria-label="input text"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-1"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              type="text"
              value=""
            />
          </div>
          <footer
            class="pf-c-modal-box__footer"
          >
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-primary"
              data-action="confirm"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Save
            </button>
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link"
              data-action="cancel"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Cancel
            </button>
          </footer>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TextInputModal render should render title 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-4"
          aria-label="input text - modal"
          aria-labelledby="pf-modal-part-2 pf-modal-part-3"
          aria-modal="true"
          class="pf-c-modal-box ins-c-inventory__detail--edit pf-m-sm"
          data-ouia-component-id="OUIA-Generated-Modal-small-3"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-2"
          role="dialog"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Modal-small-3-ModalBoxCloseButton"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          <header
            class="pf-c-modal-box__header"
          >
            <h1
              class="pf-c-modal-box__title"
              id="pf-modal-part-3"
            >
              
              <span
                class="pf-c-modal-box__title-text"
              >
                Some title
              </span>
            </h1>
          </header>
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-4"
          >
            <input
              aria-invalid="false"
              aria-label="input text"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-2"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              type="text"
              value=""
            />
          </div>
          <footer
            class="pf-c-modal-box__footer"
          >
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-primary"
              data-action="confirm"
              data-ouia-component-id="OUIA-Generated-Button-primary-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Save
            </button>
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link"
              data-action="cancel"
              data-ouia-component-id="OUIA-Generated-Button-link-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Cancel
            </button>
          </footer>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TextInputModal render should render without any props 1`] = `
<body>
  <div />
  <div />
</body>
`;
